icu4c/source/common/locid.cpp

```diff
@@ -477,6 +483,30 @@ Locale &Locale::operator=(const Locale &other)
     return *this;
 }
 
+Locale& Locale::operator=(Locale&& other) U_NOEXCEPT {
+    setToBogus();
```

Comment:
Optional: This does more than we need. Just release the buffers (and set them to fullNameBuffer so it's ok if this==&other) if heap-allocated? Or keep this call for simplicity?

Reply:
@sffc must answer this question, it was he who requested that I call `setToBogus()` instead of just trying to do the bare minimum, see commit df650e5.

Reply:
I disagree. Copy and move have quite different semantics. I want move to be fast, we are allowed to leave the source object in an undefined (but harmless) state, and we should take advantage of that.

Another difference is that copy can fail (out of memory) so an early setToBogus() simplifies error handling later on; but move never fails.

For how much work we do preparing "this" Locale object, I prefer avoiding extra work but calling a common function would be nice. We could make one like UnicodeString::releaseArray() (but plural for Locale) that does just that (maybe not even resets the pointers) and is called by the destructor, setToBogus(), and move assign. (Maybe there are more places that release the arrays.)

Reply:
Both copy and move want to be vast.  I want to be able to look at the two functions and reason about their correctness together.  I want to see move reassign ownership of a buffer where copy clones the buffer.

The fact that copy can fail is a good reason to put this->setToBogus() there but not in move.

The very act of calling a function could be more overhead than the function implementation itself.  I can see an argument either way on whether to put the array freeing here or in a helper function.